src/node/hiding.rs

```diff
@@ -0,0 +1,226 @@
+use crate::node::{
+    CoreConstructible, DisconnectConstructible, JetConstructible, WitnessConstructible,
+};
+use crate::types::{Context, Error};
+use crate::{Cmr, FailEntropy, Merkelized, Word};
+
+/// Wrapper for a node type that allows the use of hidden nodes
+/// during the construction of an expression.
+///
+/// A hidden node is useful when the CMR of a given subexpression is known,
+/// but not the expression itself.
+/// In this case, the constructor [`Hiding::hidden`] can be used.
+///
+/// Besides that,
+/// [`Hiding`] implements the [`crate::node::Constructible`] trait,
+/// which brings the usual methods for program construction.
+///
+/// ## Propagation of hidden nodes
+///
+/// Hidden nodes propagate upwards in the expression tree,
+/// towards the next `case` combinator:
+///
+/// 1. `case` with a left hidden child and a right non-hidden child becomes `assertr`.
+/// 2. `case` with a left non-hidden child and a right hidden child becomes `assertl`.
+/// 3. If `case` has two non-hidden children, then it stays `case`.
+/// 4. If `case` has two hidden children, then it itself becomes hidden.
+///
+/// If there is no `case` ancestor, then the root node becomes hidden.
+/// If the root node of a program is hidden, then the program is invalid.
+///
+/// ## Retrieval of final program
+///
+/// When the program is fully constructed,
+/// then it can be retrieved via the [`Hiding::get_node`] method.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Hiding<N> {
+    result: HidingResult<N>,
+    ctx: Context,
+}
+
+type HidingResult<N> = Result<N, Cmr>;
+
+impl<N> Hiding<N> {
+    /// Create a hidden node with the given CMR.
+    pub const fn hidden(cmr: Cmr, ctx: Context) -> Self {
+        Self {
+            result: Err(cmr),
+            ctx,
+        }
+    }
+
+    fn hidden_cloned_ctx(&self, cmr: Cmr) -> Self {
+        Self {
+            result: Err(cmr),
+            ctx: self.ctx.shallow_clone(),
+        }
+    }
+
+    /// Access the non-hidden node contained in the wrapper.
+    ///
+    /// Return `None` if the wrapper contains a hidden node.
+    pub const fn as_node(&self) -> Option<&N> {
+        match &self.result {
```

Comment:
In 60166205dcb0a46cacba79c3306a8b30505a38b8:

This function can be `self.result.as_ref().ok()`.